File: 백준/Gold/1005. ACM Craft/ACM Craft.py
from collections import deque
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
def make_build():
    n, k = map(int, input().split())
    buildings = [[]for _ in range(n+1)]
    res_time = [0 for _ in range(n+1)]
    times = list(map(int, input().split()))
    times.insert(0, 0)
    cnt = [0 for _ in range(n+1)]
    q = deque()
    for _ in range(k):
        start, end = map(int, input().split())
        buildings[start].append(end)
        cnt[end]+=1
    w = int(input())
    for i in range(1, n+1):
        if cnt[i] == 0:
            q.append(i)
            res_time[i] += times[i]
    while q:
        now = q.popleft()
        for v in buildings[now]:
            cnt[v] -= 1
            res_time[v] = max(res_time[v], res_time[now] + times[v])
            if cnt[v] == 0:
                q.append(v)
    return res_time[w]

tc = int(input())
for _ in range(tc):
    print(make_build())

File: 백준/Gold/1005. ACM Craft/ACM Craft.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

type Queue struct {
	Node []int
}

func (q *Queue) Enqueue(n int) {
	q.Node = append(q.Node, n)
}

func (q *Queue) Dequeue() (ret int) {
	if len(q.Node) <= 0 {
		return -1
	} else {
		ret = q.Node[0]
		if len(q.Node) > 1 {
			q.Node = q.Node[1:]
		} else {
			q.Node = []int{}
		}
	}
	return ret
}

func (q *Queue) Len() int {
	return len(q.Node)
}
func makeBuilding(n, w int, cnt []int, times []float64, buildings [][]int) float64 {
	var resTime = make([]float64, n+1)
	var now int
	var q = new(Queue)
	for i, v := range cnt {
		if v == 0 {
			q.Enqueue(i)
			resTime[i] += times[i]
		}
	}
	for q.Len() > 0 {
		now = q.Dequeue()
		for _, v := range buildings[now] {
			cnt[v]--
			resTime[v] = math.Max(resTime[v], resTime[now]+times[v])
			if cnt[v] == 0 {
				q.Enqueue(v)
			}
		}
	}
	return resTime[w]
}
func main() {
	reader := bufio.NewReader(os.Stdin)
	var tc, n, k, start, end, w int
	fmt.Fscanf(reader, "%d\n", &tc)
	for i := 0; i < tc; i++ {
		fmt.Fscanf(reader, "%d %d\n", &n, &k)
		var buildings = make([][]int, n+1)
		var times = make([]float64, n+1)
		var cnt = make([]int, n+1)
		for j := 1; j < n+1; j++ {
			buildings[i] = []int{}
			fmt.Fscanf(reader, "%f ", &times[j])
		}
		for j := 0; j < k; j++ {
			fmt.Fscanf(reader, "%d %d\n", &start, &end)
			buildings[start] = append(buildings[start], end)
			cnt[end]++
		}
		fmt.Fscanf(reader, "%d\n", &w)
		fmt.Printf("%0.f\n", makeBuilding(n, w, cnt, times, buildings))
	}
}
